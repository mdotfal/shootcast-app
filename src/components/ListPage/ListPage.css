@import url('https://fonts.googleapis.com/css2?family=Noto+Sans+JP:wght@400;500&display=swap');

/* List Styles */

.list {
  font-family: 'Noto Sans JP', sans-serif;
  font-size: 12px;
}

.list-cities {
  display: inline;
  height: 275px;
  overflow: auto;
}

.list-cities ul {
  display: flex;
  flex-direction: row;
  list-style-type: none;
  color:black;
  height: 55px;
  overflow: auto;
}

.list-cities li {
  display: flex;
  justify-content: center;
  align-items: center;
  padding: 10px 5px;
  text-align: center;
  background: #F6EDE7;
  box-shadow: 0 1px 2px 0;
  border-radius: 25px;
  margin: 5px 5px 10px 5px;
  width: 150px;
}

.list-cities p {
  margin-bottom: 15px;
}

/* Item Styles */

.item {
  height: 30px;
  width: 100px;
}

.city-item {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  height: 30px;
}

.city-name {
  display: flex;
  justify-content: center;
  align-items: center;
}

/* Button Styles */

button {
  margin: 10px;
  width: 20px;
  border-radius: 25px;
  border: 0;
  padding: 5px;
}

.btn {
  border-radius: 50%;
  padding: 1px;
  margin: 10px 5px 10px 15px;
  width: 20px;
}

/* Miscellaneous */

.list-forecast {
  margin: 50px 0;
}

.addcity-form {
  margin-bottom: 15px;
}

.weather-display {
  text-align: center;
  padding: 5px;
}

.forecast-items {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  overflow: auto;
  margin: auto;
}

.weather-display-items {
  margin-left: 10px;
}

@media all and ( min-width: 600px ) {

  .sidebar-home {
    width: 800px;
  }

  .list {
    margin: auto;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  .list-cities {
    min-height: 400px;
  }
  
  .list-cities ul {
    display: flex;
    flex-flow: row wrap;
    width: 800px;
    overflow: visible;
  }

}