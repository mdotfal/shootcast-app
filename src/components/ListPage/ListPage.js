import React, { Component } from 'react';
import AddCity from '../AddCity/AddCity';
import CurrentWeather from '../CurrentWeather/CurrentWeather';
import Item from '../Item/Item';
import WeatherDisplay from '../WeatherDisplay/WeatherDisplay';
import config from '../../config';
import AppContext from '../../AppContext';
import PropTypes from 'prop-types';
import './ListPage.css';
class ListPage extends Component {

  static contextType = AppContext;

  state = {
    forecastData: [],
    weatherData: [],
    city: "",
    background: "",
    isError: false,
  }

  componentDidUpdate = ( prevProps, prevState ) => {
    if( prevState.city !== this.state.city ) {
       this.fetchWeather( this.state.city );
    }
  }

  // Main FETCH calls for Weather API

  fetchWeather = () => {
    const city = !undefined ? this.state.city : "San Bruno";
    const baseForecast = `https://api.openweathermap.org/data/2.5/forecast?q=`;
    const baseWeather = `https://api.openweathermap.org/data/2.5/weather?q=`;
    const api_key = `${ config.API_KEY }`;
    const imperial = `&units=imperial`;
    const forecastUrl = `${ baseForecast }${ city }&appid=${ api_key }${ imperial }`;
    const weatherUrl = `${ baseWeather }${ city }&appid=${ api_key }${ imperial }`;

    Promise.all([
      fetch( forecastUrl ),
      fetch( weatherUrl )
    ]) 
    .then( ([ forecastRes, weatherRes ]) => {
      if( !forecastRes.ok ) {
        return forecastRes.json().then( e => Promise.reject( e ));
      }
      if( !weatherRes.ok ) {
        return weatherRes.json().then( e => Promise.reject( e ));
      }
      return Promise.all([ forecastRes.json(), weatherRes.json() ]);
    })
    .then( ([ forecastData, weatherData ]) => {
      const forecast = forecastData.list.filter( fData => fData.dt_txt.includes("18:00:00") );
      this.setState({
        isError: false,
        forecastData: forecast,
        weatherData,
      });
    })
    .catch( err => {
      this.setState({
        isError: true
      });
    });
    
  }

  // Map WeatherDisplay Component
  formatForecast = () => {
    return this.state.forecastData.map( ( data, i ) => 
      <WeatherDisplay data={ data } key={ i } /> 
    );
  }

  // Grabs cityName and adds it to state for onClick Weather fetch call
  getNameOnClick = ( cityName ) => {
    this.setState({ city: cityName });
  }

  render() {
    return (
      <div className="list">
        <AddCity />
        <div className="list-cities">
          <p>&#8678; Click a city to view Weather &#8680;</p>
          <ul>

            { this.props.cities.map( (city , i ) =>
                <li key={ i }>
                  <Item
                    id={ city.id }
                    value={ city.name }
                    city={ city }
                    getNameOnClick={ this.getNameOnClick }
                    history={ this.props.history }
                  />
                </li>)
            }
          </ul>
        </div>
        <div className="list-forecast">

          {/*  This is Weather Display */}
          <div className="city-name">
            {
              this.state.isError === true
                ? ""
                : <CurrentWeather 
                    wData={ this.state.weatherData }
                    city={ this.state.city } 
                  />
            }
          </div>
          <div className="forecast-items">
            
            {  }
            { this.state.isError === false && this.state.forecastData !== []
                ? this.formatForecast()
                : <div className="forecast-welcome">
                  <h3 style={{ color: 'red' }}>City not found!</h3>
                  <p>Check spelling and add a new city!</p>
                </div>
            }
            
          </div>
        </div>
     </div>
    );
  }
};

ListPage.propTypes = {
  cities: PropTypes.array,
};

export default ListPage;