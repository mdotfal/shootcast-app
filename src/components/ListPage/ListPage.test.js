import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ListPage from './ListPage';

describe( `ListPage component`, () => {
  it(Â `renders without errors`, () => {
    const cities = [
      { name: "San Francisco", list_id: 1 },
      { name: "Castro Valley", list_id: 2 }
    ]
    const wrapper = shallow( <ListPage cities={ cities } /> )
    expect( toJson( wrapper )).toMatchSnapshot()
  })
})