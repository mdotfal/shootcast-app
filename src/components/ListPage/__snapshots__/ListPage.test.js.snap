// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPage component renders without errors 1`] = `
<div
  className="list"
>
  <AddCity />
  <div
    className="list-cities"
  >
    <p>
      ⇦ Click a city to view Weather ⇨
    </p>
    <ul>
      <li
        key="0"
      >
        <Item
          city={
            Object {
              "list_id": 1,
              "name": "San Francisco",
            }
          }
          getNameOnClick={[Function]}
          value="San Francisco"
        />
      </li>
      <li
        key="1"
      >
        <Item
          city={
            Object {
              "list_id": 2,
              "name": "Castro Valley",
            }
          }
          getNameOnClick={[Function]}
          value="Castro Valley"
        />
      </li>
    </ul>
  </div>
  <div
    className="list-forecast"
  >
    <div
      className="city-name"
    >
      <CurrentWeather
        wData={Array []}
      />
    </div>
    <div
      className="forecast-items"
    >
      <div
        className="forecast-welcome"
      >
        <h2>
          Welcome to ShootCast!
        </h2>
        <p>
          Click a city to get weather!
        </p>
      </div>
    </div>
  </div>
</div>
`;
